typing/typemod.ml

```diff
@@ -3296,28 +3296,31 @@ let report_error ~loc _env = function
   | Cannot_hide_id Illegal_shadowing
       { shadowed_item_kind; shadowed_item_id; shadowed_item_loc;
         shadower_id; user_id; user_kind; user_loc } ->
+      let shadowed = Printtyp.human_id ~index:1 shadowed_item_id in
+      let shadower = Printtyp.human_id ~index:0 shadower_id in
```

Comment:
The hardcoded number won't be ambiguous to the user because the error message explain where it comes from. But could an other error message bind `t/2` to something else ?
`Printtyp.ident` seems to be more careful at picking an index. Perhaps it could be used here with `shadowed_item_id` added to the Env ?

Reply:
Each error message comes with its own context and `Env.t`, which is signaled to the user through the location of the error message. In other words, a name like `t/<n>` is only bound inside the context of a specific error message.

Moreover those indices are the indices that we would get if we computed the environment at the end of the definition of the shadower item, excluding ephemeral item definitions: at the end of the definition of the shadower item, it is the most recent item with its name. Then, the shadowed item is necessarily the previous non-ephemeral component sharing this name: if there were another non-ephemeral component between the shadower and shadowed itself, we would have reported this definition has an illegal shadowing before the current error.

Nevertheless, I propose to amend that behavior to also count ephemeral components:
```
include struct type t = A end
let x  = A
open struct type t = int end
type t
```
currently refers to the shadowed `t` as `t/2` when we probably want to name `t/3` which is done in #11916 in https://github.com/ocaml/ocaml/pull/11916/commits/f47ea0ccfd55af7532c216c3669f43991954e2c8 but I didn't propagate the change because that I forgot the issue with intermediary ephemeral component definitions.

Reply:
Thanks for the explanation! Would it make sense to make `Printtyp.ident` aware of the `erased_idents` of https://github.com/ocaml/ocaml/commit/f47ea0ccfd55af7532c216c3669f43991954e2c8 and use that instead of `Printtyp.human_id` ?
It seems that it was supposed to handle such cases without asking for an arbitrary index. `human_id` could be left un-exposed.

Reply:
The issue of using `Printtyp.ident` directly is that for `single illegal include`s is that we don't have an environment at hand at the point of the shadower identifier. We could re-add the erased identifiers to the printing environment of the error but I am not sure if building an artificial environment is more readable than building a specialized identifier function. It might be the case than the environement is the best option because it require less knowledge of the printing machinery of `Pprintyp`, I will think about it.

Reply:
After more investigations, using a `Printtyp.namespaced_ident` function with the error environment works well enough once the "usual" interaction between the global state of `Printtyp` and the delayed printing of `Location.msg` is correctly handled.